### PR TITLE
Fix print precision and match numpy behavior

### LINKED ISSUE #12578: Printing: mismatch result with numpy in pytorch 0.4
I noticed a weird behavior when using torch.linspace.  Please take a look into the following code, where I compare the output of pytorch and numpy for different number of steps:

```python
import torch
import numpy as np
import sys

mult_factor=0.00002


# Torch
print('11 steps')
n_ = torch.linspace(-5.0, 5.0,steps=11)* mult_factor
print(n_)

n_ = np.linspace(-5.0, 5.0,11)* mult_factor
print(n_)
print('.....')

print('12 steps')
n_ = torch.linspace(-5.0, 5.0,steps=12)* mult_factor
print(n_)

n_ = np.linspace(-5.0, 5.0,12)* mult_factor
print(n_)
print('.....')

print('13 steps')
n_ = torch.linspace(-5.0, 5.0,steps=13)* mult_factor
print(n_)

n_ = np.linspace(-5.0, 5.0,13)* mult_factor
print(n_)
print('.....')


print('14 steps')
n_ = torch.linspace(-5.0, 5.0,steps=14)* mult_factor
print(n_)

n_ = np.linspace(-5.0, 5.0,14)* mult_factor
print(n_)
```
If I run the code above I have the following output:
```
11 steps
tensor([-0.0001, -0.0001, -0.0001, -0.0000, -0.0000,  0.0000,  0.0000,  0.0000,
         0.0001,  0.0001,  0.0001])
[-1.e-04 -8.e-05 -6.e-05 -4.e-05 -2.e-05  0.e+00  2.e-05  4.e-05  6.e-05
  8.e-05  1.e-04]
.....
12 steps
tensor([-1.0000e-04, -8.1818e-05, -6.3636e-05, -4.5455e-05, -2.7273e-05,
        -9.0909e-06,  9.0909e-06,  2.7273e-05,  4.5455e-05,  6.3636e-05,
         8.1818e-05,  1.0000e-04])
[-1.00000000e-04 -8.18181818e-05 -6.36363636e-05 -4.54545455e-05
 -2.72727273e-05 -9.09090909e-06  9.09090909e-06  2.72727273e-05
  4.54545455e-05  6.36363636e-05  8.18181818e-05  1.00000000e-04]
.....
13 steps
tensor([-0.0001, -0.0001, -0.0001, -0.0000, -0.0000, -0.0000,  0.0000,  0.0000,
         0.0000,  0.0000,  0.0001,  0.0001,  0.0001])
[-1.00000000e-04 -8.33333333e-05 -6.66666667e-05 -5.00000000e-05
 -3.33333333e-05 -1.66666667e-05  0.00000000e+00  1.66666667e-05
  3.33333333e-05  5.00000000e-05  6.66666667e-05  8.33333333e-05
  1.00000000e-04]
.....
14 steps
tensor([-1.0000e-04, -8.4615e-05, -6.9231e-05, -5.3846e-05, -3.8462e-05,
        -2.3077e-05, -7.6923e-06,  7.6923e-06,  2.3077e-05,  3.8462e-05,
         5.3846e-05,  6.9231e-05,  8.4615e-05,  1.0000e-04])
[-1.00000000e-04 -8.46153846e-05 -6.92307692e-05 -5.38461538e-05
 -3.84615385e-05 -2.30769231e-05 -7.69230769e-06  7.69230769e-06
  2.30769231e-05  3.84615385e-05  5.38461538e-05  6.92307692e-05
  8.46153846e-05  1.00000000e-04]
```
Curiously, when I set an even "step" everything seems to work, if I set an odd number of step it seems there is a lack of resolution.  Note that this doesn't happen in numpy. Moreover, I tried with pytorch 0.3 and the issue doesn't arise.  I  noticed the same behavior with torch.range and torch.arange. 

What could be the problem?

### PR DESCRIPTION
**Pull request by ailzhang**
Fixes #12578 #9395.

* Fix and simplify print logic

* Follow numpy print rule https://github.com/numpy/numpy/blob/eb2bd11870731ea19a0eee72e616c7deb00f6c54/numpy/core/arrayprint.py#L859
> scientific notation is used when absolute value of the smallest number is < 1e-4 or maximum > 1e8 or the ratio of the maximum absolute value to the minimum is > 1e3

I hope I didn't break anything since there seems to be a lot of edge cases here... Here are some easy sanity checks. 
```
# For int tensor, we just print, never use scientific.
In [5]: torch.tensor(1)
Out[5]: tensor(1)
Out[2]: array(1) # numpy

In [6]: torch.tensor(10)
Out[6]: tensor(10)
Out[3]: array(10) # numpy

In [8]: torch.tensor(99000000)
Out[8]: tensor(99000000)
Out[5]: array(99000000) # numpy

In [9]: torch.tensor(100000000)
Out[9]: tensor(100000000)
Out[6]: array(100000000) # numpy

In [10]: torch.tensor(100000001)
Out[10]: tensor(100000001)
Out[7]: array(100000001) # numpy

In [11]: torch.tensor(1000000000)
Out[11]: tensor(1000000000)
Out[8]: array(1000000000) # numpy

In [12]: torch.tensor([1, 1000])
Out[12]: tensor([   1, 1000])
Out[9]: array([   1, 1000]) # numpy

In [13]: torch.tensor([1, 1010])
Out[13]: tensor([   1, 1010])
Out[10]: array([   1, 1010]) # numpy
```
For floating points, we use scientific when `max/min > 1000 || max > 1e8 || min < 1e-4`
Lines with "old" are old behaviors that either has precision issue, or not aligned with numpy
```
In [14]: torch.tensor(0.01)
Out[14]: tensor(0.0100)
Out[11]: array(0.01) # numpy

In [15]: torch.tensor(0.1)
Out[15]: tensor(0.1000)
Out[12]: array(0.1) # numpy

In [16]: torch.tensor(0.0001)
Out[16]: tensor(0.0001)
Out[14]: array(0.0001) # numpy

In [17]: torch.tensor(0.00002)
Out[17]: tensor(2.0000e-05)
Out[15]: array(2e-05) # numpy
Out[5]: tensor(0.0000) # old

In [18]: torch.tensor(1e8)
Out[18]: tensor(100000000.)
Out[16]: array(100000000.0) # numpy

In [19]: torch.tensor(1.1e8)
Out[19]: tensor(1.1000e+08)
Out[17]: array(1.1e8) # numpy 1.14.5, In <= 1.13 this was not using scientific print
Out[10]: tensor(110000000.) # old

In [20]: torch.tensor([0.01, 10.])
Out[20]: tensor([ 0.0100, 10.0000])
Out[18]: array([  0.01,  10.  ]) # numpy

In [21]: torch.tensor([0.01, 11.])
Out[21]: tensor([1.0000e-02, 1.1000e+01])
Out[19]: array([  1.00000000e-02,   1.10000000e+01]) # numpy
Out[7]: tensor([ 0.0100, 11.0000]) # old
```
When print floating number in int mode, we still need to respect rules to use scientific mode first
```
In [22]: torch.tensor([1., 1000.])
Out[22]: tensor([   1., 1000.])
Out[20]: array([    1.,  1000.]) # numpy

In [23]: torch.tensor([1., 1010.])
Out[23]: tensor([1.0000e+00, 1.0100e+03])
Out[21]: array([  1.00000000e+00,   1.01000000e+03]) # numpy
Out[9]: tensor([   1., 1010.]) # old
```

### CODE REVIEW

**Review by soumith — CHANGES_REQUESTED** [review]
the test cases in the issue description, where you match numpy behavior,, make them test cases.

File: torch/_tensor_str.py [hunk 1]
```diff
@@ -72,65 +72,52 @@ def __init__(self, tensor):
         self.sci_mode = False
         self.max_width = 1
 
+        # use tensor_view for 0-dim tensor iteration
+        tensor_view = tensor.view(tensor.nelement())
```
Comment (vishwakftw) [inline torch/_tensor_str.py]:
```suggestion
        with torch.no_grad():
            tensor_view = tensor.view(tensor.nelement())
```

I think this is better, otherwise we might end up having to track autograd graph unnecessarily IIUC.

**Comment by ailzhang** [comment]
@soumith I added a few tests. There are one thing left todo:
As you can see from examples above, Numpy does automatic trimming and padding  while we use a fixed precision of 4.  Numpy default is 8. But this causes a problem when we try to print a really long tensor, we actually force the print precision to be 4:
```
>>> np.array([123456789.])
array([1.23456789e+08])
>>> torch.tensor([123456789.])
tensor([1.2346e+08])
```
I actually think we should port the dragon4_scientific code from Numpy here to make the print prettier.  But it should be in a separate PR and depends on the priority. I opened #12797 for it. 
https://github.com/numpy/numpy/blob/f36d2d4d3f622f7901e3d5ade13e04fc05062948/numpy/core/src/multiarray/multiarraymodule.c#L3530

**Review by soumith — APPROVED** [review]
lgtm. Instead of separate expect files per each string, it feels like it's better to inline the expected strings in the tests. Making a separate file doesn't add a lot of value here, and one has to jump 1 extra file to find out what the expected value is supposed to be.

**Comment by ezyang** [comment]
I implemented inline expect tests in my spare time. Let me put it in PyTorch.
